tests/graceful-recovery.md

```diff
@@ -0,0 +1,97 @@
+# Graceful recovery from restarts
+
+## Description
+When one of the containers in the NGF Pod fails, the Gateway API configuration
+should be reapplied to the new container so that when either container
+fails, NGF should be able to recover without any assistance.
+
+## Goal
+Ensure that NGF can recover gracefully from container failures without any intervention.
+
+## Cluster Details
+
+- GKE 1.27.3-gke.100
+- us-central1-c
+- Machine type of node is e2-medium
+- 3 nodes
+
+## Setup
+
+1. Setup GKE Cluster.
+2. Clone the repo and change into the nginx-gateway-fabric directory.
+3. Check out the latest tag (unless you are installing the edge version from the main branch).
+4. Go into `deploy/manifests/nginx-gateway.yaml` and change `runAsNonRoot` from `true` to `false`.
+5. Follow the [installation instructions](https://github.com/nginxinc/nginx-gateway-fabric/blob/main/docs/installation.md)
+to deploy NGINX Gateway Fabric.
+6. In a separate terminal track NGF logs by running `kubectl -n nginx-gateway logs -f deploy/nginx-gateway`
+7. In a separate terminal track nginx container logs by running
+`kubectl -n nginx-gateway logs -f <NGF_POD> -c nginx`
+8. Exec into the nginx container inside of the NGF pod by running
+`kubectl exec -it -n nginx-gateway <NGF_POD> --container nginx -- bin/sh`
+9. Inside the nginx container, navigate to `/etc/nginx/conf.d` and ensure that
+`http.conf` and `config-version.conf` look correct.
+10. In a different terminal, deploy the
+[https-termination example](https://github.com/nginxinc/nginx-gateway-fabric/tree/main/examples/https-termination).
+11. Inside the nginx container, check `http.conf` and `config-version.config` to see
+if the configuration and version were correctly updated.
+12. Send traffic through the example application and ensure it is working correctly.
+
+## Testing when nginx-gateway container restarts
+
+1. Ensure NGF and nginx container logs are set up and traffic flows through the example application correctly.
+2. Insert ephemeral container in NGF Pod and kill the nginx-gateway process.
+    1. `kubectl debug -it -n nginx-gateway <NGF_POD> --image=busybox:1.28 --target=nginx-gateway`
+    2. run `ps -A`
+    3. run `kill <nginx-gateway_PID>` (Command should start with `/usr/bin/gateway`)
```

Comment:
by default, kill sends SIGTERM, which is for normal (graceful) termination. We need to kill the process abruptly. For that, we need use SIGKILL, like `kill -s SIGKILL`

Reply:
Do you think it would be worth having a test which checks if normal (graceful) termination works alongside SIGKILL? Do you think the test is going to miss anything important by testing through SIGKILL?

Reply:
The most important is SIGKILL, because it simulates an abprupt termination of either container.

SIGTERM is important in the context of termination, when the container will not be restarted, but deleted. Graceful termination will be tested here, because it is important for zero downtime upgrades https://github.com/nginxinc/nginx-gateway-fabric/issues/950 --  the terminating pod will need to stop accepting any new connections and finish serving any pending requests before it exits.